### DISCUSSION TITLE
How do you add a local library to a pixi project?

### DISCUSSION THREAD

**Discussion by matthewfeickert** [prefix-dev/pixi#1071] [Q&A] [answered]
While I understand that `pixi` is project focused currently rather than library development focused, it is not uncommon in my field to have a small project specific library that lives in the project repository to provide helper functions and additional functionality. How would I add such a local library to a `pixi` project?

For example, if we take the test library that lives in https://github.com/scientific-python/upload-nightly-action and try to build an example `pixi` project things fail with the following

```console
$ pixi --version
pixi 0.17.1
$ git clone https://github.com/scientific-python/upload-nightly-action.git
$ cd upload-nightly-action/
$ git reset --hard 535c72c70a11e52673bee0471be1ca52161a4b5
$ cd tests/test_package/
$ tree .
.
├── pyproject.toml
├── README.md
└── src
    └── test_package
        └── __init__.py

2 directories, 3 files
$ pixi init .
✔ Initialized project in /tmp/upload-nightly-action/tests/test_package/.
$ pixi add .
  × could not determine any available versions for . on linux-64. Either the package could not be found or version constraints on other dependencies result in a conflict.
  ╰─▶ Cannot solve the request because of: No candidates were found for . *.
      

```

I'm not sure if PR https://github.com/prefix-dev/pixi/pull/985 helps on this at all.

**Comment by ruben-arts** [comment, marked as answer]
He @matthewfeickert,

You're just in time! #985 would indeed do the first part of that. But #1044 is really going to change this experience with editable installs. It doesn't include an `pixi add x` command, yet, but we do provide the underlying functionality to add `test_package  = { path = ".", editable = true}` to your `pixi.toml`.

This way you can install unreleased, local packages.

I'm not sure how you get your `pixi` installed but to test the latest `main` you can use `cargo install --locked --git https://github.com/prefix-dev/pixi.git` or `brew install --HEAD pixi`.
